### PR TITLE
add option to train 1000-speaker subset

### PR DESCRIPTION
**Pull request by ruizhilijhu**
add option to train 1000-speaker subset: jsalt18e2e/asr1: run.sh --subset_num_spk 1000

### CODE REVIEW

File: egs/jsalt18e2e/asr1/run.sh [hunk 1]
```diff
@@ -178,21 +195,21 @@ if [ ${stage} -le 1 ]; then
...
-        /export/b{13,14,15,16}/${USER}/espnet-data/egs/jsalt18e2e/asr1/dump/${train_dev}/delta${do_delta}/storage \
+        /export/b{01,02,03,04}/${USER}/espnet-data/egs/jsalt18e2e/asr1/dump/${train_dev}/delta${do_delta}/storage \
         ${feat_dt_dir}/storage
     fi
-    dump.sh --cmd "$train_cmd" --nj 40 --do_delta $do_delta \
-        data/${train_set}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/train ${feat_tr_dir}
-    dump.sh --cmd "$train_cmd" --nj 40 --do_delta $do_delta \
-        data/${train_dev}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/dev ${feat_dt_dir}
+    [ ! -d ${feat_tr_dir}/feats.scp ] && dump.sh --cmd "$train_cmd" --nj 40 --do_delta $do_delta \
+        data/${train_set}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/train_${train_set} ${feat_tr_dir}
```
Comment (sw005320) [inline egs/jsalt18e2e/asr1/run.sh]: `train_${train_set}` -> `${train_set}`

File: egs/jsalt18e2e/asr1/run.sh [hunk 2]
```diff
@@ -178,21 +195,21 @@ if [ ${stage} -le 1 ]; then
...
+    [ ! -d ${feat_dt_dir}/feats.scp ] && dump.sh --cmd "$train_cmd" --nj 40 --do_delta $do_delta \
+        data/${train_dev}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/dev_${train_dev} ${feat_dt_dir}
```
Comment (sw005320) [inline egs/jsalt18e2e/asr1/run.sh]: `dev_${train_dev}` -> `${train_dev}`

File: egs/jsalt18e2e/asr1/run.sh [hunk 3]
```diff
@@ -91,6 +91,7 @@ subset_num_spk=""
 
 # data set
 train_set=tr_babel10${subset_num_spk:+_${subset_num_spk}spk}
+langdir=data/lang_1char${subset_num_spk:+_${subset_num_spk}spk}
```
Comment (ruizhilijhu) [inline egs/jsalt18e2e/asr1/run.sh]: Each subset experiment specifies a different lang dir

File: egs/jsalt18e2e/asr1/run.sh [hunk 4]
```diff
@@ -167,6 +175,15 @@ if [ ${stage} -le 1 ]; then
     utils/combine_data.sh data/${train_set}_org data/tr_babel_cantonese data/tr_babel_bengali data/tr_babel_pashto data/tr_babel_turkish data/tr_babel_vietnamese data/tr_babel_haitian data/tr_babel_tamil data/tr_babel_kurmanji data/tr_babel_tokpisin data/tr_babel_georgian
     utils/combine_data.sh data/${train_dev}_org data/dt_babel_cantonese data/dt_babel_bengali data/dt_babel_pashto data/dt_babel_turkish data/dt_babel_vietnamese data/dt_babel_haitian data/dt_babel_tamil data/dt_babel_kurmanji data/dt_babel_tokpisin data/dt_babel_georgian
 
+    if [ ! -z $subset_num_spk ]; then
+        # create a trainng subset with ${subset_num_spk} speakers (in total 7470)
```
[the remark this answers is not in the document]
Comment (ruizhilijhu) [reply egs/jsalt18e2e/asr1/run.sh]: Done

File: egs/jsalt18e2e/asr1/run.sh [hunk 5]
```diff
@@ -168,20 +168,19 @@ if [ ${stage} -le 0 ]; then
     done
 fi
 
-feat_tr_dir=${dumpdir}/${train_set}/delta${do_delta}; mkdir -p ${feat_tr_dir}
-feat_dt_dir=${dumpdir}/${train_dev}/delta${do_delta}; mkdir -p ${feat_dt_dir}
+feat_tr_dir=${dumpdir}/${train_set}/delta${do_delta}; mkdir -p ${feat_tr_dir}/json_${train_set}
+feat_dt_dir=${dumpdir}/${train_dev}/delta${do_delta}; mkdir -p ${feat_dt_dir}/json_${train_set}
```
Comment (ruizhilijhu) [inline egs/jsalt18e2e/asr1/run.sh]: For different subset experiment, the generated data.json is different for all train/test datasets due to the different lang dir.

Comment (sw005320) [reply egs/jsalt18e2e/asr1/run.sh]: We don't need lang dir.
Also is `/json_${train_set}` required?

My concern is that with subset, we actually need to prepare the dev and test set directories depending on the ${train_set} as it depend on CMN obtained by ${train_set}. 
How about 
`feat_dt_dir=${dumpdir}/${train_dev}_${train_set}/delta${do_delta}`
`feat_recog_dir=${dumpdir}/${rtask}_${train_set}/delta${do_delta}`

File: egs/jsalt18e2e/asr1/run.sh [hunk 6]
```diff
@@ -178,52 +194,52 @@ if [ ${stage} -le 1 ]; then
...
-    dump.sh --cmd "$train_cmd" --nj 40 --do_delta $do_delta \
-        data/${train_dev}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/dev ${feat_dt_dir}
+    [ ! -d ${feat_tr_dir}/feats.scp ] && dump.sh --cmd "$train_cmd" --nj 40 --do_delta $do_delta \
+        data/${train_set}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/${train_set} ${feat_tr_dir}
+    [ ! -d ${feat_dt_dir}/feats.scp ] && dump.sh --cmd "$train_cmd" --nj 40 --do_delta $do_delta \
+        data/${train_dev}/feats.scp data/${train_set}/cmvn.ark exp/dump_feats/${train_dev} ${feat_dt_dir}
    for rtask in ${recog_set}; do
-        feat_recog_dir=${dumpdir}/${rtask}/delta${do_delta}; mkdir -p ${feat_recog_dir}
-        dump.sh --cmd "$train_cmd" --nj 40 --do_delta $do_delta \
+        feat_recog_dir=${dumpdir}/${rtask}/delta${do_delta}; mkdir -p ${feat_recog_dir}/json_${train_set}
```
Comment (sw005320) [inline egs/jsalt18e2e/asr1/run.sh]: is `/json_${train_set}` required?

File: egs/jsalt18e2e/asr1/run.sh (hunk 5 above)
Comment (ruizhilijhu) [reply egs/jsalt18e2e/asr1/run.sh]: We use only one language dictionary trained from the whole set other than forming language dictionary from subset?

Comment (sw005320) [reply egs/jsalt18e2e/asr1/run.sh]: Yes, and I think we may not have to make it depending on ${train_set}, as we always use the language dictionary trained from the whole set.

Comment (ruizhilijhu) [reply egs/jsalt18e2e/asr1/run.sh]: Gotcha, that makes sense. Please find the new commit for the update.

**Comment by ruizhilijhu** [comment]
Fixed the issues

**Pull request merged**